### PR TITLE
Allow empty column names in CSVWithNames/TSVWithNames

### LINKED ISSUE #47458: Allow empty column names in CSVWithNames/TSVWithNames
Sophia Air Quality Dataset

Download: https://www.kaggle.com/datasets/hmavrodiev/sofia-air-quality-dataset
License: https://opendatacommons.org/licenses/odbl/1-0/

`tar xvf sofia-air-quality-dataset.tar.zst`

CSV files have the following header:
```
,sensor_id,location,lat,lon,timestamp,pressure,temperature,humidity
```

And while DESCRIBE works:
```
milovidov-desktop :) DESC file('*.csv', CSVWithNames)

DESCRIBE TABLE file('*.csv', CSVWithNames)

Query id: c8a4145f-d42d-46db-8612-535b50f1bb5d

┌─name────────┬─type────────────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│             │ Nullable(Int64)         │              │                    │         │                  │                │
│ sensor_id   │ Nullable(Int64)         │              │                    │         │                  │                │
│ location    │ Nullable(Int64)         │              │                    │         │                  │                │
│ lat         │ Nullable(Float64)       │              │                    │         │                  │                │
│ lon         │ Nullable(Float64)       │              │                    │         │                  │                │
│ timestamp   │ Nullable(DateTime64(9)) │              │                    │         │                  │                │
│ pressure    │ Nullable(Float64)       │              │                    │         │                  │                │
│ temperature │ Nullable(Float64)       │              │                    │         │                  │                │
│ humidity    │ Nullable(Float64)       │              │                    │         │                  │                │
└─────────────┴─────────────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘
```

Parsing them yields an error:
```
milovidov-desktop :) SELECT * FROM file('*.csv', CSVWithNames)

SELECT *
FROM file('*.csv', CSVWithNames)

Query id: d9b19aa6-a2c1-4e87-b759-1a6f7cb184a5


0 rows in set. Elapsed: 0.173 sec. 

Received exception:
Code: 42. DB::Exception: Expected not empty name: While processing ``: While processing SELECT ``, sensor_id, location, lat, lon, timestamp, pressure, temperature, humidity FROM file('*.csv', 'CSVWithNames'). (NUMBER_OF_ARGUMENTS_DOESNT_MATCH)
```

And it does not recognize the format as CSVWithNames by default.

### Proposed solution

Option 1: Use `c1` or the first unused `cN` as the column name instead of empty.
Option 2: Ignore empty columns from the file.

### PR DESCRIPTION
**Pull request by iiiuwioajdks**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
When your csv file contains empty columns, like：
```

┌─name────────┬─type────────────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│             │ Nullable(Int64)         │              │                    │         │                  │                │
│ sensor_id   │ Nullable(Int64)         │              │                    │         │                  │                │
│ location    │ Nullable(Int64)         │              │                    │         │                  │                │
│ lat         │ Nullable(Float64)       │              │                    │         │                  │                │
│ lon         │ Nullable(Float64)       │              │                    │         │                  │                │
│ timestamp   │ Nullable(DateTime64(9)) │              │                    │         │                  │                │
│ pressure    │ Nullable(Float64)       │              │                    │         │                  │                │
│ temperature │ Nullable(Float64)       │              │                    │         │                  │                │
│ humidity    │ Nullable(Float64)       │              │                    │         │                  │                │
└─────────────┴─────────────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘
```
If you execute the following query:
```
SELECT * FROM file('*.csv', CSVWithNames)
```
An exception is thrown before the modification, after which the empty columns are removed, producing the following output columns:
```
┌─sensor_id─┬─location─┬───────────────lat─┬─────────────────lon─┬─── ────────────────────timestamp─┬─pressure─┬─temperature─┬─────────────humidity─┐
```

close #47458

### CODE REVIEW

**Comment by iiiuwioajdks** [comment]
The style check seems to be failing, do I need to fix this? @Avogar

**Comment by Avogar** [comment]
Thank you for your contribution!
Removing empty column names in TreeRewriter is a possible solution, but not the best. Empty column name can appear only during automatic schema inference from data format, so, it will be better to add this logic in `readSchemaFromFormat` function in `src/Formats/ReadSchemaUtils.cpp` file.

File: src/Formats/ReadSchemaUtils.cpp [hunk 1]
```diff
@@ -191,20 +199,29 @@ ColumnsDescription readSchemaFromFormat(
...
-                        "{} file format doesn't support schema inference. You must specify the structure manually",
-                        format_name);
-
+        throw Exception(
+            ErrorCodes::BAD_ARGUMENTS,
+            "{} file format doesn't support schema inference. You must specify the structure manually",
+            format_name);
+    names_and_types.erase(
+        std::remove_if(names_and_types.begin(), names_and_types.end(), [](const NameAndTypePair & pair) { return pair.name.empty(); }),
+        names_and_types.end());
```
Comment (Avogar) [inline src/Formats/ReadSchemaUtils.cpp:208]: Let's write a comment why we remove columns with empty names

**Review by Avogar — APPROVED** [review]

**Comment by robot-ch-test-poll4** [comment]
This is an automated comment for commit 5ada38550238e20f2be19b2a26f7b1d3543ea9b6 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/47496/5ada38550238e20f2be19b2a26f7b1d3543ea9b6/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

**Pull request merged**